File: enroll.py
import base64
import json
import cv2
import face_recognition
import mysql.connector
import numpy as np

DB_CONFIG = {
    "host": "127.0.0.1",
    "port": 5522,  # Uses the forwarded local port
    "user": "",  # Replace with your DB user
    "password": "",  # Replace with your DB password
    "database": "",
}




def capture_and_enroll(pupil_id):
    """Opens local webcam, captures face vector using dlib, and updates MySQL."""
    video_capture = cv2.VideoCapture(0)

    print("\nLook at the camera. Press 's' to capture and save, or 'q' to quit.")

    while True:
        ret, frame = video_capture.read()
        if not ret:
            print("Failed to access camera.")
            break

        # Show live feed with instructions
        cv2.putText(
            frame,
            "Press 'S' to Save Vector | 'Q' to Quit",
            (20, 30),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.7,
            (0, 255, 0),
            2,
        )
        cv2.imshow("Enroll Pupil Face", frame)

        key = cv2.waitKey(1) & 0xFF

        if key == ord("s"):
            # Convert BGR to RGB
            rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)

            # Detect faces and compute dlib 128-d encoding
            # boxes = face_recognition.face_locations(rgb_frame)
            # ✅ Upsample 2 or 3 times to detect faces under shadows / angles
            #boxes = face_recognition.face_locations(rgb_frame, number_of_times_to_upsample=2)
            # Uses dlib's Convolutional Neural Network detector instead of HOG
            boxes = face_recognition.face_locations(rgb_frame, model="cnn")
            encodings = face_recognition.face_encodings(rgb_frame, boxes)

            if len(encodings) == 1:
                # Convert numpy array vector to JSON string
                vector_json = json.dumps(encodings[0].tolist())

                # Save to MySQL
                try:
                    conn = mysql.connector.connect(**DB_CONFIG)
                    cursor = conn.cursor()

                    query = """
                        UPDATE Pupils 
                        SET face_vector = %s, has_vector = 1 
                        WHERE ID = %s
                    """
                    cursor.execute(query, (vector_json, pupil_id))
                    conn.commit()

                    cursor.close()
                    conn.close()

                    print(
                        f" Successfully saved face vector for Pupil ID {pupil_id}!"
                    )
                    break
                except mysql.connector.Error as err:
                    print(f"Database error: {err}")
            elif len(encodings) == 0:
                print("⚠️ No face detected! Please look directly at the camera.")
            else:
                print("⚠️ Multiple faces detected! Ensure only one person is in frame.")

        elif key == ord("q"):
            print("Enrollment canceled.")
            break

    video_capture.release()
    cv2.destroyAllWindows()


if __name__ == "__main__":
    target_pupil_id = input("Enter Pupil ID to enroll: ").strip()
    if target_pupil_id.isdigit():
        capture_and_enroll(int(target_pupil_id))
    else:
        print("Invalid Pupil ID.")

File: enroll.php
<?php
// Database configuration
$host     = 'localhost';
$db_name  = '';
$username = 'YOUR_DB_USER';                 // Replace with your DB username
$password = 'YOUR_DB_PASSWORD';             // Replace with your DB password

try {
    $pdo = new PDO("mysql:host=$host;dbname=$db_name;charset=utf8mb4", $username, $password);
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
} catch (PDOException $e) {
    die("Database Connection Failed: " . $e->getMessage());
}

$message = '';

// Handle Form Submission (Update Vector)
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['update_vector'])) {
    $pupil_id    = !empty($_POST['pupil_id']) ? (int)$_POST['pupil_id'] : 0;
    $face_vector = trim($_POST['face_vector'] ?? '');

    if ($pupil_id <= 0 || empty($face_vector)) {
        $message = '<div class="alert alert-danger">Please select a pupil and capture a face vector first.</div>';
    } else {
        // Automatically determine Primary Key column name
        $id_col = 'PupilID'; // Change to 'ID' or 'id' if needed

        $sql = "UPDATE Pupils SET face_vector = :face_vector, has_vector = 1 WHERE {$id_col} = :pupil_id";
        $stmt = $pdo->prepare($sql);
        $executed = $stmt->execute([
            ':face_vector' => $face_vector,
            ':pupil_id'    => $pupil_id
        ]);

        if ($executed) {
            $message = '<div class="alert alert-success">Face vector updated successfully!</div>';
        } else {
            $message = '<div class="alert alert-danger">Failed to update pupil record.</div>';
        }
    }
}

// Fetch pupils
$pupils = [];
try {
    $stmt = $pdo->query("SELECT * FROM Pupils ORDER BY Surname, Forename");
    $pupils = $stmt->fetchAll(PDO::FETCH_ASSOC);
} catch (PDOException $e) {
    $message = '<div class="alert alert-danger">SQL Query Failed: ' . htmlspecialchars($e->getMessage()) . '</div>';
}
?>

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Enroll Face Vector</title>
    <!-- Bootstrap CSS -->
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.0/dist/css/bootstrap.min.css" rel="stylesheet">
    <!-- Select2 CSS -->
    <link href="https://cdn.jsdelivr.net/npm/select2@4.1.0-rc.0/dist/css/select2.min.css" rel="stylesheet" />
    <link href="https://cdn.jsdelivr.net/npm/select2-bootstrap-5-theme@1.3.0/dist/select2-bootstrap-5-theme.min.css" rel="stylesheet" />

    <!-- face-api.js Library -->
    <script defer src="https://cdn.jsdelivr.net/npm/@vladmandic/face-api/dist/face-api.js"></script>

    <style>
        #video-container {
            position: relative;
            width: 100%;
            max-width: 480px;
            margin: 0 auto;
            background: #000;
            border-radius: 8px;
            overflow: hidden;
        }
        video {
            width: 100%;
            height: auto;
            display: block;
        }
    </style>
</head>
<body class="bg-light">

<div class="container mt-4 mb-5">
    <div class="row justify-content-center">
        <div class="col-md-8">
            <div class="card shadow-sm">
                <div class="card-header bg-primary text-white">
                    <h4 class="card-title mb-0">Enroll Pupil Face Vector</h4>
                </div>
                <div class="card-body">
                    <?= $message; ?>

                    <form method="POST" action="enroll.php">
                        <input type="hidden" name="update_vector" value="1">

                        <!-- Select Pupil -->
                        <div class="mb-4">
                            <label for="pupil_id" class="form-label fw-bold">1. Select Pupil *</label>
                            <select class="form-select select2" id="pupil_id" name="pupil_id" required>
                                <option value="" disabled selected>Search by name, year, or house...</option>
                                <?php foreach ($pupils as $pupil): 
                                    $id = $pupil['PupilID'] ?? $pupil['ID'] ?? $pupil['id'] ?? 0;
                                    $hasVector = !empty($pupil['has_vector']);
                                ?>
                                    <option value="<?= htmlspecialchars($id); ?>">
                                        <?= htmlspecialchars(($pupil['Surname'] ?? '') . ', ' . ($pupil['Forename'] ?? '')); ?> 
                                        (Year: <?= htmlspecialchars($pupil['Year_Group'] ?? 'N/A'); ?>, House: <?= htmlspecialchars($pupil['House'] ?? 'N/A'); ?>)
                                        <?= $hasVector ? ' [Vector Enrolled]' : ' [No Vector]'; ?>
                                    </option>
                                <?php endforeach; ?>
                            </select>
                        </div>

                        <!-- Webcam Section -->
                        <div class="mb-4 text-center">
                            <label class="form-label fw-bold d-block">2. Capture Face Vector</label>
                            
                            <div id="video-container" class="mb-3">
                                <video id="webcam" autoplay muted playsinline></video>
                            </div>

                            <div id="status-badge" class="badge bg-secondary mb-3">Loading face detection models...</div>

                            <div>
                                <button type="button" id="btn-capture" class="btn btn-success" disabled>
                                    📸 Capture Face & Generate Vector
                                </button>
                            </div>
                        </div>

                        <!-- Hidden / Readonly Vector Data Container -->
                        <div class="mb-4">
                            <label for="face_vector" class="form-label fw-bold">Generated Vector Output</label>
                            <textarea class="form-control font-monospace" id="face_vector" name="face_vector" rows="3" readonly placeholder="Vector coordinates will appear here after clicking capture..." required></textarea>
                        </div>

                        <div class="d-grid">
                            <button type="submit" id="btn-submit" class="btn btn-primary btn-lg" disabled>Save Vector to Database</button>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    </div>
</div>

<!-- Scripts -->
<script src="https://code.jquery.com/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/select2@4.1.0-rc.0/dist/js/select2.min.js"></script>

<script>
$(document).ready(function() {
    $('.select2').select2({ theme: 'bootstrap-5' });
});

const video = document.getElementById('webcam');
const statusBadge = document.getElementById('status-badge');
const btnCapture = document.getElementById('btn-capture');
const btnSubmit = document.getElementById('btn-submit');
const vectorInput = document.getElementById('face_vector');

// Load Face-API models from CDN
async function loadModels() {
    const MODEL_URL = 'https://cdn.jsdelivr.net/npm/@vladmandic/face-api/model/';
    
    try {
        await faceapi.nets.ssdMobilenetv1.loadFromUri(MODEL_URL);
        await faceapi.nets.faceLandmark68Net.loadFromUri(MODEL_URL);
        await faceapi.nets.faceRecognitionNet.loadFromUri(MODEL_URL);
        
        statusBadge.className = 'badge bg-info';
        statusBadge.innerText = 'Models loaded. Starting webcam...';
        startWebcam();
    } catch (err) {
        statusBadge.className = 'badge bg-danger';
        statusBadge.innerText = 'Error loading models: ' + err.message;
    }
}

// Start Video Stream
function startWebcam() {
    navigator.mediaDevices.getUserMedia({ video: {} })
        .then(stream => {
            video.srcObject = stream;
            statusBadge.className = 'badge bg-success';
            statusBadge.innerText = 'Ready to capture';
            btnCapture.disabled = false;
        })
        .catch(err => {
            statusBadge.className = 'badge bg-danger';
            statusBadge.innerText = 'Webcam access denied or unavailable';
        });
}

// Capture and Generate Vector
btnCapture.addEventListener('click', async () => {
    statusBadge.className = 'badge bg-warning text-dark';
    statusBadge.innerText = 'Detecting face & calculating descriptors...';

    const detection = await faceapi.detectSingleFace(video)
                                   .withFaceLandmarks()
                                   .withFaceDescriptor();

    if (detection) {
        // Convert Float32Array vector descriptor to JSON array string
        const vectorArray = Array.from(detection.descriptor);
        vectorInput.value = JSON.stringify(vectorArray);

        statusBadge.className = 'badge bg-success';
        statusBadge.innerText = 'Face captured successfully!';
        btnSubmit.disabled = false;
    } else {
        statusBadge.className = 'badge bg-danger';
        statusBadge.innerText = 'No face detected! Ensure good lighting and look directly at the camera.';
    }
});

// Initialize on page load
window.addEventListener('DOMContentLoaded', loadModels);
</script>

</body>
</html>
